File: day2.lua

shape_scores = {
	['rock'] = 1,
	['paper'] = 2,
	['scissors'] = 3,
}

outcome_scores = {
	['win'] = 6,
	['lose'] = 0,
	['draw'] = 3,
}

shapes = {
	['A'] = 'rock',
	['B'] = 'paper',
	['C'] = 'scissors',
	['X'] = 'rock',
	['Y'] = 'paper',
	['Z'] = 'scissors'
}

p2_outcomes = {
	['X'] = 'lose',
	['Y'] = 'draw',
	['Z'] = 'win'
}

function calculate_score_p1(shapecode1, shapecode2)
	shape1 = shapes[shapecode1]
	shape2 = shapes[shapecode2]

	score = shape_scores[shape2]
	
	if shape1 == shape2 then
		score = score + outcome_scores.draw
	elseif 
		(shape2 == 'rock' and shape1 == 'scissors')
		or  
		(shape2 == 'paper' and shape1 == 'rock')
		or  
		(shape2 == 'scissors' and shape1 == 'paper') then
		score = score + outcome_scores.win
	else
		score = score + outcome_scores.lose
	end
	return score
end

function calculate_score_p2(shapecode1,outcomecode)
	score = 0
	shape1 = shapes[shapecode1]
	shape2 = ''
	outcome = p2_outcomes[outcomecode]
	if outcome == 'draw' then
		shape2 = shape1
	elseif outcome == 'win' then
		if shape1 == 'rock' then shape2 = 'paper' end
		if shape1 == 'paper' then shape2 = 'scissors' end
		if shape1 == 'scissors' then shape2 = 'rock' end
	else
		if shape1 == 'rock' then shape2 = 'scissors' end
		if shape1 == 'paper' then shape2 = 'rock' end
		if shape1 == 'scissors' then shape2 = 'paper' end
	end
	score = score + outcome_scores[outcome] + shape_scores[shape2]
	return score
end

total_p1 = 0
total_p2 = 0

for line in io.lines('day2.txt') do
	for opponent,response in string.gmatch(line,'([%w])%s(%w)') do
		total_p1 = total_p1 + calculate_score_p1(opponent,response)
		total_p2 = total_p2 + calculate_score_p2(opponent,response)
	end
end

-- day 2 part 1 
print(total_p1)

-- day 2 part 1 
print(total_p2)

File: day2.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>

enum shape{ROCK=1,PAPER,SCISSORS};
enum outcome{WIN=6,LOSE=0,DRAW=3};

int part2_outcome(char code){
        switch(code){
                case 'X':return LOSE;
                case 'Y':return DRAW;
                case 'Z':return WIN;
        }
}

int getShape(char shapecode){
        if(shapecode == 'A' || shapecode == 'X'){return ROCK;}
        if(shapecode == 'B' || shapecode == 'Y'){return PAPER;}
        return SCISSORS;
}

int calculate_score_p1(char shapecode1, char shapecode2){
        int shape1 = getShape(shapecode1);
        int shape2 = getShape(shapecode2);
        int score = shape2;
        if(shape1 == shape2)
                score+=DRAW;
        else if(
                shape2 == ROCK && shape1 == SCISSORS ||
                shape2 == PAPER && shape1 == ROCK ||
                shape2 == SCISSORS && shape1 == PAPER
        )
                score+= WIN;
        else
                score+=LOSE;
        return score;
}

int calculate_score_p2(char shapecode1,char outcomecode){
        int shape2 = 0;
        int shape1 = getShape(shapecode1);
        int outcome = part2_outcome(outcomecode);
        if(outcome == DRAW) shape2=shape1;
        else if (outcome == WIN){
                if(shape1 == ROCK) shape2 = PAPER;
                if(shape1 == PAPER) shape2 = SCISSORS;
                if(shape1 == SCISSORS) shape2 += ROCK;
        }else{
                if(shape1 == ROCK) shape2 = SCISSORS;
                if(shape1 == PAPER) shape2 = ROCK;
                if(shape1 == SCISSORS) shape2 = PAPER;
        }
        return outcome+shape2;
}

int main(){
        int score_p1 = 0,score_p2=0;
        char p1,p2;
        FILE* input = fopen("day2.txt","r");
        while(!feof(input)){
                fscanf(input,"%c %c\n",&p1,&p2);
                score_p1 += calculate_score_p1(p1,p2);
                score_p2 += calculate_score_p2(p1,p2);
        }
        fclose(input);
        printf("Part 1: %d\n",score_p1);
        printf("Part 2: %d\n",score_p2);
        return 0;
}
